File: lib/java/src/main/java/uapi/internal/generation/GenerateRandomStruct.java
package uapi.internal.generation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

import uapi.internal.types.UFieldDeclaration;
import uapi.internal.types.UTypeDeclaration;

public class GenerateRandomStruct {
    public static Object generateRandomStruct(
            Map<String, UFieldDeclaration> referenceStruct, GenerateContext ctx) {
        final var startingStruct = ctx.useBlueprintValue ? (Map<String, Object>) ctx.blueprintValue : new HashMap<>();

        final var sortedReferenceStruct = new ArrayList<>(referenceStruct.entrySet());
        Collections.sort(sortedReferenceStruct, (e1, e2) -> e1.getKey().compareTo(e2.getKey()));

        final var obj = new TreeMap<String, Object>();
        for (final var field : sortedReferenceStruct) {
            final var fieldName = field.getKey();
            final var fieldDeclaration = field.getValue();
            final var blueprintValue = startingStruct.get(fieldName);
            final var useBlueprintValue = startingStruct.containsKey(fieldName);
            final UTypeDeclaration typeDeclaration = fieldDeclaration.typeDeclaration;

            final Object value;
            if (useBlueprintValue) {
                value = typeDeclaration.generateRandomValue(
                        ctx.copyWithNewBlueprintValueAndUseBlueprintValue(blueprintValue, useBlueprintValue));
            } else {
                if (!fieldDeclaration.optional) {
                    value = typeDeclaration
                            .generateRandomValue(ctx.copyWithNewBlueprintValueAndUseBlueprintValue(null, false));
                } else {
                    if (!ctx.includeOptionalFields
                            || (ctx.randomizeOptionalFields && ctx.randomGenerator.nextBoolean())) {
                        continue;
                    }
                    value = typeDeclaration
                            .generateRandomValue(ctx.copyWithNewBlueprintValueAndUseBlueprintValue(null, false));
                }
            }

            obj.put(fieldName, value);
        }
        return obj;
    }
}

File: lib/py/uapi/internal/generation/GenerateRandomStruct.py
from typing import TYPE_CHECKING, cast


if TYPE_CHECKING:
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.types.UFieldDeclaration import UFieldDeclaration
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
    from uapi.internal.generation.GenerateContext import GenerateContext


def generate_random_struct(blueprint_value: object, use_blueprint_value: bool,
                           reference_struct: dict[str, 'UFieldDeclaration'], ctx: 'GenerateContext') -> object:
    sorted_reference_struct = sorted(
        reference_struct.items(), key=lambda x: x[0])

    starting_struct = cast(
        dict[str, object], blueprint_value) if use_blueprint_value else {}

    obj = {}
    for field_name, field_declaration in sorted_reference_struct:
        this_use_blueprint_value = field_name in starting_struct
        type_declaration = field_declaration.type_declaration

        if this_use_blueprint_value:
            this_blueprint_value = starting_struct.get(field_name)
            value = type_declaration.generate_random_value(
                this_blueprint_value, this_use_blueprint_value, ctx)
        else:
            if not field_declaration.optional:
                value = type_declaration.generate_random_value(
                    None, False, ctx)
            else:
                if not ctx.include_optional_fields or (ctx.randomize_optional_fields and ctx.random_generator.next_boolean()):
                    continue
                value = type_declaration.generate_random_value(
                    None, False, ctx)

        obj[field_name] = value

    return obj
